File: src-tauri/src/commands/ai.rs
use tauri::{AppHandle, State};
use crate::state::AppState;
use crate::models::ai::*;
use crate::error::AppError;

#[tauri::command]
pub async fn ai_chat(
    state: State<'_, AppState>,
    request: AiChatRequest,
) -> Result<AiChatResponse, AppError> {
    state.ai_service.chat(request).await
}

#[tauri::command]
pub async fn ai_chat_stream(
    app: AppHandle,
    state: State<'_, AppState>,
    request: AiChatRequest,
) -> Result<(), AppError> {
    state.ai_service.chat_stream(&app, request).await
}

#[tauri::command]
pub fn get_ai_functions() -> Vec<FunctionInfo> {
    crate::services::function_call::get_function_infos()
}

#[derive(serde::Serialize)]
pub struct FunctionInfo {
    pub name: String,
    pub description: String,
}

File: src/services/tauri/ai.ts
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import type { AiChatRequest, AiChatResponse } from "@/types/chat";
import type { Todo } from "@/types/todo";

export const aiService = {
  async chat(request: AiChatRequest): Promise<AiChatResponse<Todo>> {
    return invoke("ai_chat", { request }) as Promise<AiChatResponse<Todo>>;
  },

  async chatStream(
    request: AiChatRequest,
    onChunk: (content: string) => void,
    onFunctionCall?: (name: string, result: unknown) => void,
  ): Promise<void> {
    const unlisten = await listen<{
      content?: string;
      function_call?: { name: string; result: unknown };
    }>("ai-stream-chunk", (event) => {
      if (event.payload.content) {
        onChunk(event.payload.content);
      }
      if (event.payload.function_call && onFunctionCall) {
        onFunctionCall(
          event.payload.function_call.name,
          event.payload.function_call.result,
        );
      }
    });

    try {
      await invoke("ai_chat_stream", { request });
    } finally {
      unlisten();
    }
  },
};

